File: local/python/frost_rf_local.py
#!/usr/bin/env python
"""Portable local Random Forest workflow for the three frost endpoints.

Input tables must contain the predictor names listed in the feature manifest and
the endpoint columns. This reference implementation is independent of the HPC
scheduler and works with CSV or Parquet files on Windows, macOS and Linux.
"""

from __future__ import annotations

import argparse
import json
from datetime import datetime, timezone
from pathlib import Path

import joblib
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.impute import SimpleImputer
from sklearn.metrics import (
    average_precision_score,
    balanced_accuracy_score,
    brier_score_loss,
    f1_score,
    mean_absolute_error,
    mean_squared_error,
    precision_score,
    r2_score,
    recall_score,
    roc_auc_score,
)
from sklearn.model_selection import GroupKFold, KFold, cross_val_predict
from sklearn.pipeline import Pipeline


ENDPOINTS = {
    "probability": {"column": "frost_any", "kind": "classifier"},
    "frost_days": {"column": "frost_days", "kind": "poisson_regressor"},
    "seasonal_tmin_c": {"column": "observed_season_tmin_c", "kind": "regressor"},
}


def read_table(path: Path) -> pd.DataFrame:
    suffix = path.suffix.lower()
    if suffix == ".csv":
        return pd.read_csv(path)
    if suffix in {".parquet", ".pq"}:
        return pd.read_parquet(path)
    raise ValueError(f"Unsupported table format: {path}")


def write_table(frame: pd.DataFrame, path: Path) -> None:
    if path.suffix.lower() == ".csv":
        frame.to_csv(path, index=False)
    else:
        frame.to_parquet(path, index=False)


def load_features(path: Path) -> list[str]:
    manifest = pd.read_csv(path)
    if "feature" not in manifest:
        raise ValueError("Feature manifest must contain a 'feature' column")
    features = manifest["feature"].dropna().astype(str).tolist()
    if not features or len(features) != len(set(features)):
        raise ValueError("Feature manifest is empty or contains duplicates")
    return features


def estimator(kind: str, trees: int, jobs: int, seed: int) -> Pipeline:
    common = dict(n_estimators=trees, min_samples_leaf=4, max_features=0.45,
                  n_jobs=jobs, random_state=seed)
    if kind == "classifier":
        model = RandomForestClassifier(class_weight="balanced_subsample", **common)
    elif kind == "poisson_regressor":
        model = RandomForestRegressor(criterion="poisson", **common)
    else:
        model = RandomForestRegressor(criterion="squared_error", **common)
    return Pipeline([("imputer", SimpleImputer(strategy="median")), ("model", model)])


def split_strategy(frame: pd.DataFrame, group_column: str | None, folds: int, seed: int):
    if group_column and group_column in frame and frame[group_column].nunique() >= folds:
        return GroupKFold(n_splits=folds), frame[group_column].astype(str).to_numpy(), "grouped"
    return KFold(n_splits=folds, shuffle=True, random_state=seed), None, "random"


def classification_metrics(y: np.ndarray, probability: np.ndarray) -> dict:
    prediction = probability >= 0.5
    return {
        "ROC_AUC": roc_auc_score(y, probability),
        "PR_AUC": average_precision_score(y, probability),
        "balanced_accuracy": balanced_accuracy_score(y, prediction),
        "sensitivity": recall_score(y, prediction, zero_division=0),
        "precision": precision_score(y, prediction, zero_division=0),
        "F1": f1_score(y, prediction, zero_division=0),
        "Brier_score": brier_score_loss(y, probability),
    }


def regression_metrics(y: np.ndarray, prediction: np.ndarray) -> dict:
    return {
        "RMSE": mean_squared_error(y, prediction) ** 0.5,
        "MAE": mean_absolute_error(y, prediction),
        "R2": r2_score(y, prediction),
    }


def train(args: argparse.Namespace) -> int:
    table = read_table(args.training)
    features = load_features(args.features)
    missing = sorted(set(features) - set(table.columns))
    if missing:
        raise ValueError(f"Training table is missing {len(missing)} predictors; first: {missing[:10]}")
    if args.quick_check and len(table) > args.quick_check:
        table = table.sample(args.quick_check, random_state=args.seed).reset_index(drop=True)

    args.output.mkdir(parents=True, exist_ok=True)
    models, metrics = {}, []
    for index, (endpoint, spec) in enumerate(ENDPOINTS.items()):
        target = spec["column"]
        if target not in table:
            raise ValueError(f"Missing endpoint column: {target}")
        work = table.loc[table[target].notna()].reset_index(drop=True)
        x = work[features]
        y = work[target].to_numpy()
        cv, groups, validation = split_strategy(work, args.group_column, args.folds, args.seed)
        pipe = estimator(spec["kind"], args.trees, args.jobs, args.seed + index)
        method = "predict_proba" if spec["kind"] == "classifier" else "predict"
        cv_prediction = cross_val_predict(pipe, x, y, cv=cv, groups=groups, method=method,
                                          n_jobs=1)
        if spec["kind"] == "classifier":
            cv_prediction = cv_prediction[:, 1]
            score = classification_metrics(y.astype(int), cv_prediction)
        else:
            score = regression_metrics(y.astype(float), cv_prediction.astype(float))
        metrics.append({"endpoint": endpoint, "n": len(work), "validation": validation, **score})
        pipe.fit(x, y)
        models[endpoint] = pipe
        print(f"LOCAL_RF_ENDPOINT_OK={endpoint} n={len(work)} validation={validation}", flush=True)

    bundle = {
        "status": "LOCAL_RF_THREE_ENDPOINTS_OK",
        "trained_at": datetime.now(timezone.utc).isoformat(),
        "features": features,
        "endpoints": ENDPOINTS,
        "models": models,
        "training_rows": len(table),
        "random_seed": args.seed,
    }
    joblib.dump(bundle, args.output / "frost_rf_three_endpoints.joblib", compress=3)
    pd.DataFrame(metrics).to_csv(args.output / "validation_metrics.csv", index=False)
    contract = {k: v for k, v in bundle.items() if k != "models"}
    (args.output / "model_contract.json").write_text(json.dumps(contract, indent=2), encoding="utf-8")

    if args.prediction:
        predict_table(args.prediction, args.output / "frost_rf_three_endpoints.joblib",
                      args.output / "predictions.parquet")
    return 0


def predict_table(input_path: Path, model_path: Path, output_path: Path) -> None:
    table = read_table(input_path)
    bundle = joblib.load(model_path)
    missing = sorted(set(bundle["features"]) - set(table.columns))
    if missing:
        raise ValueError(f"Prediction table is missing {len(missing)} predictors; first: {missing[:10]}")
    x = table[bundle["features"]]
    result = table.copy()
    result["predicted_frost_probability"] = bundle["models"]["probability"].predict_proba(x)[:, 1]
    result["predicted_frost_days"] = np.maximum(0, bundle["models"]["frost_days"].predict(x))
    result["predicted_seasonal_tmin_c"] = bundle["models"]["seasonal_tmin_c"].predict(x)
    write_table(result, output_path)
    print(f"LOCAL_RF_PREDICTIONS_OK={output_path}", flush=True)


def parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("--training", type=Path, required=True)
    p.add_argument("--features", type=Path, required=True)
    p.add_argument("--output", type=Path, required=True)
    p.add_argument("--prediction", type=Path)
    p.add_argument("--group-column", default="station_id")
    p.add_argument("--trees", type=int, default=700)
    p.add_argument("--folds", type=int, default=5)
    p.add_argument("--jobs", type=int, default=-1)
    p.add_argument("--seed", type=int, default=20260807)
    p.add_argument("--quick-check", type=int, metavar="N", help="Use at most N rows for a reduced-scale integration check")
    return p


if __name__ == "__main__":
    raise SystemExit(train(parser().parse_args()))

File: local/R/frost_rf_local.R
#!/usr/bin/env Rscript
# Portable pure-R implementation of the three-endpoint frost Random Forest.

suppressPackageStartupMessages({
  library(ranger)
  library(data.table)
  library(jsonlite)
})

parse_args <- function(x) {
  out <- list()
  for (item in x) {
    if (!startsWith(item, "--") || !grepl("=", item, fixed = TRUE)) next
    pair <- strsplit(sub("^--", "", item), "=", fixed = TRUE)[[1]]
    out[[pair[1]]] <- paste(pair[-1], collapse = "=")
  }
  out
}

read_table <- function(path) {
  if (grepl("\\.parquet$", path, ignore.case = TRUE)) {
    if (!requireNamespace("arrow", quietly = TRUE)) stop("Install package 'arrow' for Parquet input")
    return(as.data.frame(arrow::read_parquet(path)))
  }
  as.data.frame(fread(path))
}

write_table <- function(x, path) {
  if (grepl("\\.parquet$", path, ignore.case = TRUE)) {
    if (!requireNamespace("arrow", quietly = TRUE)) stop("Install package 'arrow' for Parquet output")
    arrow::write_parquet(x, path)
  } else fwrite(x, path)
}

impute_fit <- function(x) vapply(x, function(v) median(v, na.rm = TRUE), numeric(1))
impute_apply <- function(x, medians) {
  for (name in names(medians)) x[[name]][is.na(x[[name]])] <- medians[[name]]
  x
}

auc_rank <- function(y, p) {
  positive <- p[y == 1]; negative <- p[y == 0]
  if (!length(positive) || !length(negative)) return(NA_real_)
  ranks <- rank(c(positive, negative), ties.method = "average")
  (sum(ranks[seq_along(positive)]) - length(positive) * (length(positive) + 1) / 2) /
    (length(positive) * length(negative))
}

pr_auc <- function(y, p) {
  ord <- order(p, decreasing = TRUE); y <- y[ord]
  tp <- cumsum(y == 1); fp <- cumsum(y == 0)
  recall <- tp / sum(y == 1); precision <- tp / pmax(tp + fp, 1)
  sum(diff(c(0, recall)) * precision)
}

classification_metrics <- function(y, p) {
  pred <- as.integer(p >= 0.5); tp <- sum(pred == 1 & y == 1); tn <- sum(pred == 0 & y == 0)
  fp <- sum(pred == 1 & y == 0); fn <- sum(pred == 0 & y == 1)
  sensitivity <- tp / max(tp + fn, 1); specificity <- tn / max(tn + fp, 1)
  precision <- tp / max(tp + fp, 1)
  data.frame(ROC_AUC = auc_rank(y, p), PR_AUC = pr_auc(y, p),
             balanced_accuracy = (sensitivity + specificity) / 2,
             sensitivity = sensitivity, specificity = specificity, precision = precision,
             F1 = 2 * precision * sensitivity / max(precision + sensitivity, 1e-12),
             Brier_score = mean((p - y)^2))
}

regression_metrics <- function(y, p) data.frame(
  RMSE = sqrt(mean((p - y)^2)), MAE = mean(abs(p - y)),
  R2 = 1 - sum((p-y)^2) / sum((y-mean(y))^2)
)

fit_model <- function(data, features, target, kind, trees, threads, seed) {
  if (kind == "classifier") data[[target]] <- factor(data[[target]], levels = c(0, 1))
  args <- list(dependent.variable.name = target, data = data[, c(features, target), drop = FALSE],
               num.trees = trees, mtry = max(1, floor(length(features) * 0.45)),
               min.node.size = 4, num.threads = threads, seed = seed,
               write.forest = TRUE, importance = "permutation")
  if (kind == "classifier") { args$probability <- TRUE; args$class.weights <- c("0"=1, "1"=1) }
  else args$splitrule <- if (kind == "poisson_regressor") "poisson" else "variance"
  do.call(ranger, args)
}

predict_endpoint <- function(model, data, kind) {
  p <- predict(model, data = data)$predictions
  if (kind == "classifier") p[, "1"] else as.numeric(p)
}

args <- parse_args(commandArgs(trailingOnly = TRUE))
required <- c("training", "features", "output")
if (any(!required %in% names(args))) stop("Required: --training=... --features=... --output=...")
dir.create(args$output, recursive = TRUE, showWarnings = FALSE)
trees <- as.integer(ifelse(is.null(args$trees), 700, args$trees))
folds <- as.integer(ifelse(is.null(args$folds), 5, args$folds))
threads <- as.integer(ifelse(is.null(args$threads), max(1, parallel::detectCores()-1), args$threads))
seed <- as.integer(ifelse(is.null(args$seed), 20260807, args$seed)); set.seed(seed)
group_column <- ifelse(is.null(args$group_column), "station_id", args$group_column)

table <- read_table(args$training)
manifest <- fread(args$features); features <- manifest$feature
missing <- setdiff(features, names(table)); if (length(missing)) stop("Missing predictors: ", paste(head(missing,10),collapse=", "))
if (!is.null(args$quick_check) && nrow(table) > as.integer(args$quick_check)) table <- table[sample(nrow(table),as.integer(args$quick_check)),]
medians <- impute_fit(table[,features,drop=FALSE]); table[,features] <- impute_apply(table[,features,drop=FALSE],medians)
endpoints <- list(probability=c("frost_any","classifier"), frost_days=c("frost_days","poisson_regressor"), seasonal_tmin_c=c("observed_season_tmin_c","regressor"))
models <- list(); metric_rows <- list()

for (i in seq_along(endpoints)) {
  key <- names(endpoints)[i]; target <- endpoints[[i]][1]; kind <- endpoints[[i]][2]
  work <- table[!is.na(table[[target]]),,drop=FALSE]
  if (group_column %in% names(work) && length(unique(work[[group_column]])) >= folds) {
    groups <- unique(work[[group_column]]); fold_map <- setNames(sample(rep(seq_len(folds),length.out=length(groups))),groups)
    fold_id <- fold_map[as.character(work[[group_column]])]; validation <- "grouped"
  } else { fold_id <- sample(rep(seq_len(folds),length.out=nrow(work))); validation <- "random" }
  cvp <- rep(NA_real_,nrow(work))
  for (fold in seq_len(folds)) {
    train <- work[fold_id != fold,,drop=FALSE]; test <- work[fold_id == fold,,drop=FALSE]
    model <- fit_model(train,features,target,kind,trees,threads,seed+i+fold)
    cvp[fold_id == fold] <- predict_endpoint(model,test[,features,drop=FALSE],kind)
  }
  score <- if (kind == "classifier") classification_metrics(as.integer(work[[target]]),cvp) else regression_metrics(as.numeric(work[[target]]),cvp)
  metric_rows[[key]] <- cbind(endpoint=key,n=nrow(work),validation=validation,score)
  models[[key]] <- fit_model(work,features,target,kind,trees,threads,seed+i)
  message("LOCAL_R_ENDPOINT_OK=",key," n=",nrow(work)," validation=",validation)
}

bundle <- list(status="LOCAL_R_RF_THREE_ENDPOINTS_OK",features=features,medians=medians,
               endpoints=endpoints,models=models,seed=seed)
saveRDS(bundle,file.path(args$output,"frost_rf_three_endpoints.rds"),compress="xz")
fwrite(rbindlist(metric_rows,fill=TRUE),file.path(args$output,"validation_metrics_R.csv"))
write(toJSON(bundle[setdiff(names(bundle),"models")],pretty=TRUE,auto_unbox=TRUE),file.path(args$output,"model_contract_R.json"))

if (!is.null(args$prediction)) {
  pred <- read_table(args$prediction); pred[,features] <- impute_apply(pred[,features,drop=FALSE],medians)
  pred$predicted_frost_probability <- predict_endpoint(models$probability,pred[,features,drop=FALSE],"classifier")
  pred$predicted_frost_days <- pmax(0,predict_endpoint(models$frost_days,pred[,features,drop=FALSE],"poisson_regressor"))
  pred$predicted_seasonal_tmin_c <- predict_endpoint(models$seasonal_tmin_c,pred[,features,drop=FALSE],"regressor")
  write_table(pred,file.path(args$output,"predictions_R.csv"))
}
message("LOCAL_R_PIPELINE_OK=",normalizePath(args$output))
